### PR TITLE
[DC] Add fork/sink materialization passes

### PR DESCRIPTION
**Pull request by mortbopet**

### CODE REVIEW

File: lib/Dialect/DC/Transforms/DCMaterialization.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  for (auto [op, forkRes] : llvm::zip(opsToProcess, newFork->getResults()))
+    replaceFirstUse(op, result, forkRes);
+}
+
+// Insert Fork Operation for every operation and function argument with more
+// than one successor.
+static LogicalResult addForkOps(Region &r, OpBuilder &rewriter) {
+  for (Operation &op : r.getOps()) {
+    // Ignore terminators, and don't add Forks to Forks.
+    if (op.getNumSuccessors() == 0 && !isa<ForkOp>(op)) {
```
Comment (RamirezLucas) [inline lib/Dialect/DC/Transforms/DCMaterialization.cpp:58]: Isn't it possible that a `ForkOp` instance that was present in the input IR before the fork/sink materialization pass runs has one of its results used multiple times? In that case, it seems we could end up with a value used more than once after the pass is finished. Does the pass assume that no forks/sinks are present in the input IR?

File: lib/Dialect/DC/Transforms/DCMaterialization.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+  return success();
+}
+
+namespace circt {
+namespace dc {
+
+// Create sink for every unused result
+LogicalResult addSinkOps(Region &r, OpBuilder &rewriter) {
+  for (Block &block : r) {
```
Comment (RamirezLucas) [inline lib/Dialect/DC/Transforms/DCMaterialization.cpp:81]: Why do we iterate over all blocks here but not in `addForkOps`? I would expect the logic to be the same.

**Review by RamirezLucas — APPROVED** [review]
LGTM (once build/tests are passing)! Just left two questions but I'm most likely missing something

File: include/circt/Dialect/DC/DCPasses.td [hunk 3]
```diff
@@ -11,4 +11,24 @@
 
 include "mlir/Pass/PassBase.td"
 
+def DCMaterializeForksSinks : InterfacePass<"dc-materialize-forks-sinks", "mlir::FunctionOpInterface"> {
+  let summary = "Materialize fork and sink operations.";
+  let description = [{
+    This pass analyses a dc.func operation and inserts fork and sink
```
Comment (teqdruid) [inline include/circt/Dialect/DC/DCPasses.td:17]: Outdated? IIRC you got rid of dc.func.

File: lib/Dialect/DC/Transforms/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  LINK_LIBS PUBLIC
+  CIRCTHW
+  CIRCTDC
+  CIRCTSupport
+  CIRCTTransforms
+  MLIRIR
+  MLIRPass
+  MLIRTransformUtils
+  MLIRMemRefDialect
```
Comment (teqdruid) [inline lib/Dialect/DC/Transforms/CMakeLists.txt:15]: Memref and affine don't seem necessary

File: lib/Dialect/DC/Transforms/DCMaterialization.cpp (hunk 1 above)
Comment (teqdruid) [reply lib/Dialect/DC/Transforms/DCMaterialization.cpp:58]: It shouldn't assume that IMO.

File: lib/Dialect/DC/Transforms/DCMaterialization.cpp (hunk 2 above)
Comment (teqdruid) [reply lib/Dialect/DC/Transforms/DCMaterialization.cpp:81]: Shouldn't `walk` be used to get to all the blocks in a design recursively?

File: lib/Dialect/DC/Transforms/DCMaterialization.cpp (hunk 1 above)
Comment (mortbopet) [reply lib/Dialect/DC/Transforms/DCMaterialization.cpp:58]: Most definitely - to create a vague excuse for myself, this code is just duplicated from the Handshake stuff. I'll fix it and add a test for it, good catch 👍

Comment (mortbopet) [reply lib/Dialect/DC/Transforms/DCMaterialization.cpp:58]: This actually exposed an issue in the fork-to-fork canonicalizer - will fix in a later comit.

**Pull request merged**